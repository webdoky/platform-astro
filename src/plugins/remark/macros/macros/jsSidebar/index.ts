import type { Root } from 'mdast';
import { EXIT } from 'unist-util-visit';

import type { AstroFile } from '../../../validate-astro-file.ts';
import createMacro from '../../create-macro.ts';

import generalLinks from './general-links.ts';
import quickLinksSections from './quick-links-sections.ts';
import referenceSections from './reference-sections.ts';
import referenceTitleLinks from './reference-title-links.ts';

function macro(tree: Root, file: AstroFile) {
  if (file.data.astro.frontmatter.sidebar) {
    console.log(
      'file.data.astro.frontmatter.sidebar',
      file.data.astro.frontmatter.sidebar,
    );
    throw new Error('Sidebar already exists');
  }
  file.data.astro.frontmatter.sidebar = [
    {
      links: generalLinks(tree, file),
      sections: quickLinksSections(tree, file),
    },
    {
      links: referenceTitleLinks(tree, file),
      sections: referenceSections(tree, file),
    },
  ];
}

const jsSidebar = createMacro(
  'jsSidebar',
  (_node, index, parent, tree, file) => {
    macro(tree, file);
    parent.children.splice(index, 1);
    return [EXIT];
  },
);

export default jsSidebar;
