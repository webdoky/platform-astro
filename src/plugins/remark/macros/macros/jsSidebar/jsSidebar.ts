import type { Root } from 'mdast';
import { EXIT, visit } from 'unist-util-visit';

import type { AstroFile } from '../../../validate-astro-file.js';
import type { AbstractMacroParentNode, MacroNode } from '../../types.js';

import generalLinks from './general-links.ts';
import quickLinksSections from './quick-links-sections.ts';
import referenceSections from './reference-sections.ts';
import referenceTitleLinks from './reference-title-links.ts';

function macro(tree: Root, file: AstroFile) {
  file.data.astro.frontmatter.sidebar = [
    {
      links: generalLinks(tree, file),
      sections: quickLinksSections(tree, file),
    },
    {
      links: referenceTitleLinks(tree, file),
      sections: referenceSections(tree, file),
    },
  ];
}

export default function jsSidebar(tree: Root, file: AstroFile) {
  visit(
    tree,
    'macro',
    (node: MacroNode, index, parent: AbstractMacroParentNode) => {
      if (node.name !== 'jsSidebar') return;
      macro(tree, file);
      if (!parent) throw new Error('Parent is undefined');
      if (index === undefined) throw new Error('Index is undefined');
      parent.children.splice(index, 1);
      return [EXIT];
    },
  );
}
