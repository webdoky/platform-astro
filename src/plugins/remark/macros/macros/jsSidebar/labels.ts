const labels = {
  JavaScript: 'JavaScript',
  Overview: 'Огляд JavaScript',
  Tutorials: 'Підручники',
  Guide: 'Посібники з JavaScript',
  Guide_Introduction: 'Вступ',
  Guide_Grammar: 'Граматика та типи',
  Guide_Control_flow: 'Керування потоком та обробка помилок',
  Guide_Loops: 'Цикли та ітерування',
  Guide_Functions: 'Функції',
  Guide_Expressions: 'Вирази та оператори',
  Guide_Numbers: 'Числа та дати',
  Guide_Text: 'Форматування тексту',
  Guide_RegExp: 'Регулярні вирази',
  Guide_Indexed_collections: 'Колекції з індексами',
  Guide_keyed_collections: 'Колекції з ключами',
  Guide_Objects: "Робота з об'єктами",
  Guide_OOP: "Докладно про об'єктну модель",
  Guide_promises: 'Використання промісів',
  Guide_Iterators_Generators: 'Ітератори та генератори',
  Guide_Meta: 'Метапрограмування',
  Guide_Modules: 'Модулі JavaScript',
  Complete_beginners: 'Для цілковитих початківців',
  Basics: 'Основи JavaScript',
  First_steps: 'Перші кроки з JavaScript',
  Building_blocks: 'Конструктивні блоки JavaScript',
  Intermediate: 'Середній рівень',
  Introducing_objects: "Вступ до об'єктів у JavaScript",
  'Client-side_APIs': 'Клієнтські API Вебу',
  Frameworks: 'Клієнтські фреймворки у JavaScript',
  'Re-introduction': 'Повторний вступ до JavaScript',
  Data_structures: 'Структури даних у JavaScript',
  Equality: 'Порівняння на схожість і однаковість',
  Closures: 'Замикання',
  Advanced: 'Поглиблене вивчення',
  Inheritance: 'Наслідування та ланцюжок прототипів',
  Strict_mode: 'Суворий режим',
  Typed_arrays: 'Типізовані масиви JavaScript',
  Memory_Management: "Керування пам'яттю",
  Event_Loop: 'Модель багатопотоковості та Цикл подій',
  Reference: 'Довідник',
  Global_Objects: "Вбудовані об'єкти",
  Operators: 'Вирази та оператори',
  Statements: 'Інструкції та оголошення',
  Functions: 'Функції',
  Classes: 'Класи',
  Errors: 'Помилки',
  More: 'Різне',
  Lexical_grammar: 'Лексична граматика',
  Enumerability: 'Перелічуваність і належність властивостей',
  Data_types: 'Типи і структури даних',
  Iteration_protocols: 'Протоколи ітерування',
  Transitioning_to_strict_mode: 'Перехід до суворого режиму',
  Template_strings: 'Шаблонні літерали',
  Deprecated_features: 'Нерекомендовані можливості',
};

export default labels;
