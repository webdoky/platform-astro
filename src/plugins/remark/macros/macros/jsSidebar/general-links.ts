import type { Root } from 'mdast';

import hasPage from '../../../../registry/has-page.ts';
import type { AstroFile } from '../../../validate-astro-file.ts';

import labels from './labels.ts';

export default function generalLinks(_tree: Root, file: AstroFile) {
  const targetLocale = process.env.TARGET_LOCALE;
  const currentPath = `/${targetLocale}/docs/${file.data.astro.frontmatter.slug}`;
  return [
    {
      slug: `Web/JavaScript`,
      title: labels.JavaScript,
    },
    {
      slug: `Web/JavaScript/Tutorials`,
      title: labels.Tutorials,
    },
  ].map(({ slug, ...otherAttributes }) => {
    const path = `/${targetLocale}/docs/${slug}`;
    return {
      ...otherAttributes,
      path: path,
      hasTranslation: hasPage(slug),
      isCurrent: path === currentPath,
    };
  });
}
