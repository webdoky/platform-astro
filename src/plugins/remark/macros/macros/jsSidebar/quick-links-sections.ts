import hasPage from '../../../../registry/has-page.ts';
import type { AstroFile } from '../../../validate-astro-file.ts';

import labels from './labels.ts';

export default function quickLinksSections(file: AstroFile) {
  const targetLocale = process.env.TARGET_LOCALE;
  const currentSlug = file.data.astro.frontmatter.slug;
  const completeBeginners = [
    {
      slug: `Learn/Getting_started_with_the_web/JavaScript_basics`,
      title: labels.Basics,
      isCurrent:
        currentSlug === 'Learn/Getting_started_with_the_web/JavaScript_basics',
    },
    {
      slug: `Learn/JavaScript/First_steps`,
      title: labels.First_steps,
      isCurrent: currentSlug === 'Learn/JavaScript/First_steps',
    },
    {
      slug: `Learn/JavaScript/Building_blocks`,
      title: labels.Building_blocks,
      isCurrent: currentSlug === 'Learn/JavaScript/Building_blocks',
    },
    {
      slug: `Learn/JavaScript/Objects`,
      title: labels.Introducing_objects,
      isCurrent: currentSlug === 'Learn/JavaScript/Objects',
    },
  ];

  const guide = [
    {
      slug: `Web/JavaScript/Guide/Introduction`,
      title: labels.Guide_Introduction,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Introduction',
    },
    {
      slug: `Web/JavaScript/Guide/Grammar_and_Types`,
      title: labels.Guide_Grammar,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Grammar_and_Types',
    },
    {
      slug: `Web/JavaScript/Guide/Control_flow_and_error_handling`,
      title: labels.Guide_Control_flow,
      isCurrent:
        currentSlug === 'Web/JavaScript/Guide/Control_flow_and_error_handling',
    },
    {
      slug: `Web/JavaScript/Guide/Loops_and_iteration`,
      title: labels.Guide_Loops,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Loops_and_iteration',
    },
    {
      slug: `Web/JavaScript/Guide/Functions`,
      title: labels.Guide_Functions,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Functions',
    },
    {
      slug: `Web/JavaScript/Guide/Expressions_and_Operators`,
      title: labels.Guide_Expressions,
      isCurrent:
        currentSlug === 'Web/JavaScript/Guide/Expressions_and_Operators',
    },
    {
      slug: `Web/JavaScript/Guide/Numbers_and_dates`,
      title: labels.Guide_Numbers,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Numbers_and_dates',
    },
    {
      slug: `Web/JavaScript/Guide/Text_formatting`,
      title: labels.Guide_Text,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Text_formatting',
    },
    {
      slug: `Web/JavaScript/Guide/Regular_Expressions`,
      title: labels.Guide_RegExp,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Regular_Expressions',
    },
    {
      slug: `Web/JavaScript/Guide/Indexed_collections`,
      title: labels.Guide_Indexed_collections,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Indexed_collections',
    },
    {
      slug: `Web/JavaScript/Guide/Keyed_collections`,
      title: labels.Guide_keyed_collections,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Keyed_collections',
    },
    {
      slug: `Web/JavaScript/Guide/Working_with_Objects`,
      title: labels.Guide_Objects,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Working_with_Objects',
    },
    {
      slug: `Web/JavaScript/Guide/Details_of_the_Object_Model`,
      title: labels.Guide_OOP,
      isCurrent:
        currentSlug === 'Web/JavaScript/Guide/Details_of_the_Object_Model',
    },
    {
      slug: `Web/JavaScript/Guide/Using_promises`,
      title: labels.Guide_promises,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Using_promises',
    },
    {
      slug: `Web/JavaScript/Guide/Iterators_and_generators`,
      title: labels.Guide_Iterators_Generators,
      isCurrent:
        currentSlug === 'Web/JavaScript/Guide/Iterators_and_generators',
    },
    {
      slug: `Web/JavaScript/Guide/Meta_programming`,
      title: labels.Guide_Meta,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Meta_programming',
    },
    {
      slug: `Web/JavaScript/Guide/Modules`,
      title: labels.Guide_Modules,
      isCurrent: currentSlug === 'Web/JavaScript/Guide/Modules',
    },
  ];

  const intermediate = [
    {
      slug: `Learn/Tools_and_testing/Client-side_JavaScript_frameworks`,
      title: labels.Frameworks,
      isCurrent:
        currentSlug ===
        'Learn/Tools_and_testing/Client-side_JavaScript_frameworks',
    },
    {
      slug: `Learn/JavaScript/Client-side_web_APIs`,
      title: labels['Client-side_APIs'],
      isCurrent: currentSlug === 'Learn/JavaScript/Client-side_web_APIs',
    },
    {
      slug: `Web/JavaScript/A_re-introduction_to_JavaScript`,
      title: labels['Re-introduction'],
      isCurrent:
        currentSlug === 'Web/JavaScript/A_re-introduction_to_JavaScript',
    },
    {
      slug: `Web/JavaScript/Data_structures`,
      title: labels.Data_structures,
      isCurrent: currentSlug === 'Web/JavaScript/Data_structures',
    },
    {
      slug: `Web/JavaScript/Equality_comparisons_and_sameness`,
      title: labels.Equality,
      isCurrent:
        currentSlug === 'Web/JavaScript/Equality_comparisons_and_sameness',
    },
    {
      slug: `Web/JavaScript/Closures`,
      title: labels.Closures,
      isCurrent: currentSlug === 'Web/JavaScript/Closures',
    },
  ];

  const advanced = [
    {
      slug: `Web/JavaScript/Inheritance_and_the_prototype_chain`,
      title: labels.Inheritance,
      isCurrent:
        currentSlug === 'Web/JavaScript/Inheritance_and_the_prototype_chain',
    },
    {
      slug: `Web/JavaScript/Reference/Strict_mode`,
      title: labels.Strict_mode,
      isCurrent: currentSlug === 'Web/JavaScript/Reference/Strict_mode',
    },
    {
      slug: `Web/JavaScript/Typed_arrays`,
      title: labels.Typed_arrays,
      isCurrent: currentSlug === 'Web/JavaScript/Typed_arrays',
    },
    {
      slug: `Web/JavaScript/Memory_Management`,
      title: labels.Memory_Management,
      isCurrent: currentSlug === 'Web/JavaScript/Memory_Management',
    },
    {
      slug: `Web/JavaScript/EventLoop`,
      title: labels.Event_Loop,
      isCurrent: currentSlug === 'Web/JavaScript/EventLoop',
    },
  ];

  return [
    {
      title: labels.Complete_beginners,
      items: completeBeginners.map((item) => ({
        ...item,
        hasTranslation: hasPage(item.slug),
        path: `/${targetLocale}/docs/${item.slug}`,
      })),
      expanded: completeBeginners.some(({ slug }) => slug === currentSlug),
    },
    {
      title: labels.Guide,
      items: guide.map((item) => ({
        ...item,
        hasTranslation: hasPage(item.slug),
        path: `/${targetLocale}/docs/${item.slug}`,
      })),
      expanded: guide.some(({ slug }) => slug === currentSlug),
    },
    {
      title: labels.Intermediate,
      items: intermediate.map((item) => ({
        ...item,
        hasTranslation: hasPage(item.slug),
        path: `/${targetLocale}/docs/${item.slug}`,
      })),
      expanded: intermediate.some(({ slug }) => slug === currentSlug),
    },
    {
      title: labels.Advanced,
      items: advanced.map((item) => ({
        ...item,
        hasTranslation: hasPage(item.slug),
        path: `/${targetLocale}/docs/${item.slug}`,
      })),
      expanded: advanced.some(({ slug }) => slug === currentSlug),
    },
  ];
}
