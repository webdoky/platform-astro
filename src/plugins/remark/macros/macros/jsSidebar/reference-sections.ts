import getAll from '../../../../registry/get-all.ts';
import hasPage from '../../../../registry/has-page.ts';
import type { RawPage } from '../../../../registry/validation.ts';
import sortBySlug from '../../../../utils/sort-by-slug.ts';
import type { AstroFile } from '../../../validate-astro-file.ts';

import labels from './labels.ts';

function sortByTitle({ title: titleA = '' }, { title: titleB = '' }) {
  return titleA.localeCompare(titleB, process.env.TARGET_LOCALE);
}

export default function referenceSections(file: AstroFile) {
  const targetLocale = process.env.TARGET_LOCALE;
  const currentPath = `/${targetLocale}/docs/${file.data.astro.frontmatter.slug}`;
  const jsPages = getAll().filter(({ slug }) =>
    slug.startsWith('Web/JavaScript/Reference/'),
  );

  const globalObjectsPages = jsPages.filter(({ slug }) =>
    slug.startsWith('Web/JavaScript/Reference/Global_Objects'),
  );
  const operatorsPages = jsPages.filter(({ slug }) =>
    slug.startsWith(`Web/JavaScript/Reference/Operators`),
  );
  const statementsPages = jsPages.filter(({ slug }) =>
    slug.startsWith(`Web/JavaScript/Reference/Statements`),
  );
  const functionsPages = jsPages.filter(({ slug }) =>
    slug.startsWith(`Web/JavaScript/Reference/Functions`),
  );
  const classesPages = jsPages.filter(({ slug }) =>
    slug.startsWith(`Web/JavaScript/Reference/Classes`),
  );
  const errorsPages = jsPages.filter(({ slug }) =>
    slug.startsWith(`Web/JavaScript/Reference/Errors`),
  );

  const morePages = [
    {
      slug: `Web/JavaScript/JavaScript_technologies_overview`,
      title: labels.Overview,
    },
    {
      slug: `Web/JavaScript/Reference/Lexical_grammar`,
      title: labels.Lexical_grammar,
    },
    {
      slug: `Web/JavaScript/Data_structures`,
      title: labels.Data_structures,
    },
    {
      slug: `Web/JavaScript/Enumerability_and_ownership_of_properties`,
      title: labels.Enumerability,
    },
    {
      slug: `Web/JavaScript/Reference/Iteration_protocols`,
      title: labels.Iteration_protocols,
    },
    {
      slug: `Web/JavaScript/Reference/Strict_mode`,
      title: labels.Strict_mode,
    },
    {
      slug: `Web/JavaScript/Reference/Strict_mode/Transitioning_to_strict_mode`,
      title: labels.Transitioning_to_strict_mode,
    },
    {
      slug: `Web/JavaScript/Reference/Template_literals`,
      title: labels.Template_strings,
    },
    {
      slug: `Web/JavaScript/Reference/Deprecated_and_obsolete_features`,
      title: labels.Deprecated_features,
    },
  ].map(({ slug, ...otherAttributes }) => ({
    ...otherAttributes,
    hasTranslation: hasPage(slug),
    path: `/${targetLocale}/docs/${slug}/`,
    isCurrent: slug === file.data.astro.frontmatter.slug,
  }));

  const pageToNavItem = ({ slug, title }: RawPage) => {
    const path = `/${targetLocale}/docs/${slug}/`;
    return {
      hasTranslation: hasPage(slug),
      isCurrent: path === currentPath,
      path: path,
      slug,
      title,
    };
  };

  return [
    {
      title: labels.Global_Objects,
      items: globalObjectsPages
        .map((item) => pageToNavItem(item))
        .sort(sortBySlug),
      expanded: globalObjectsPages.some(
        ({ slug }) => slug === file.data.astro.frontmatter.slug,
      ),
    },
    {
      title: labels.Operators,
      items: operatorsPages.map((item) => pageToNavItem(item)).sort(sortBySlug),
      expanded: operatorsPages.some(
        ({ slug }) => slug === file.data.astro.frontmatter.slug,
      ),
    },
    {
      title: labels.Statements,
      items: statementsPages
        .map((item) => pageToNavItem(item))
        .sort(sortBySlug),
      expanded: statementsPages.some(
        ({ slug }) => slug === file.data.astro.frontmatter.slug,
      ),
    },
    {
      title: labels.Functions,
      items: functionsPages.map((item) => pageToNavItem(item)).sort(sortBySlug),
      expanded: functionsPages.some(
        ({ slug }) => slug === file.data.astro.frontmatter.slug,
      ),
    },
    {
      title: labels.Classes,
      items: classesPages.map((item) => pageToNavItem(item)).sort(sortBySlug),
      expanded: classesPages.some(
        ({ slug }) => slug === file.data.astro.frontmatter.slug,
      ),
    },
    {
      title: labels.Errors,
      items: errorsPages.map((item) => pageToNavItem(item)).sort(sortBySlug),
      expanded: errorsPages.some(
        ({ slug }) => slug === file.data.astro.frontmatter.slug,
      ),
    },
    {
      title: labels.More,
      items: morePages,
      expanded: morePages.some(({ path }) => path === currentPath),
    },
  ];
}
