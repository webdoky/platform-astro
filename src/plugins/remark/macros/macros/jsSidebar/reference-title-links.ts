import type { Root } from 'mdast';

import hasPage from '../../../../registry/has-page.ts';
import type { AstroFile } from '../../../validate-astro-file.ts';

import labels from './labels.js';

export default function referenceTitleLinks(_tree: Root, file: AstroFile) {
  const targetLocale = process.env.TARGET_LOCALE;
  return [
    {
      slug: `Web/JavaScript/Reference`,
      title: labels.Reference,
    },
  ].map(({ slug, ...otherAttributes }) => ({
    ...otherAttributes,
    hasTranslation: hasPage(slug),
    path: `/${targetLocale}/docs/${slug}`,
    isCurrent: slug === file.data.astro.frontmatter.slug,
  }));
}
