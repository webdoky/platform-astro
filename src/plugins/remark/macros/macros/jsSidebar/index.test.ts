import type { Root } from 'mdast';
import { afterAll, beforeEach, describe, expect, it, vi } from 'vitest';

import type { AstroFile } from '../../../validate-astro-file.ts';
import type { AbstractMacroParentNode } from '../../types.ts';

import jsSidebar from './index.ts';

vi.mock('../../../../registry/get-all.ts', () => ({
  default: vi.fn().mockReturnValue([]),
}));
vi.mock('../../../../registry/has-page.ts', () => ({
  default: vi.fn().mockReturnValue(true),
}));

describe('jsSidebar', () => {
  const OLD_ENV = process.env;

  beforeEach(() => {
    vi.resetModules(); // Most important - it clears the cache
    process.env = {
      TARGET_LOCALE: 'uk',
    };
  });

  afterAll(() => {
    process.env = OLD_ENV; // Restore old environment
  });
  it('should throw an error if the sidebar already exists', () => {
    expect(() => {
      const tree = {
        type: 'root',
        children: [
          {
            type: 'macro',
            name: 'jsSidebar',
            parameters: [],
          },
        ],
      };
      const file = {
        data: {
          astro: {
            frontmatter: {
              sidebar: [],
              'page-type': '',
              title: '',
              slug: '',
            },
          },
        },
      } as any;
      jsSidebar(
        {
          type: 'macro',
          name: 'jsSidebar',
          parameters: [],
        },
        0,
        tree,
        tree as Root,
        file,
      );
    }).toThrowError('Sidebar already exists');
  });
  it('should add a sidebar to the frontmatter', () => {
    const tree: AbstractMacroParentNode = {
      type: 'root',
      children: [
        {
          type: 'macro',
          name: 'jsSidebar',
          parameters: [],
        },
      ],
    };
    const file = {
      data: {
        astro: {
          frontmatter: {},
        },
      },
    } as AstroFile;
    jsSidebar(
      {
        type: 'macro',
        name: 'jsSidebar',
        parameters: [],
      },
      0,
      tree,
      tree as Root,
      file,
    );
    expect(file.data.astro.frontmatter.sidebar).toMatchSnapshot();
  });
});
