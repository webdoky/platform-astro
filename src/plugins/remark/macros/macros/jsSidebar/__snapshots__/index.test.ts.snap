// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsSidebar > should add a sidebar to the frontmatter 1`] = `
[
  {
    "links": [
      {
        "hasTranslation": true,
        "isCurrent": false,
        "path": "/uk/docs/Web/JavaScript/",
        "title": "JavaScript",
      },
      {
        "hasTranslation": true,
        "isCurrent": false,
        "path": "/uk/docs/Web/JavaScript/Tutorials/",
        "title": "Підручники",
      },
    ],
    "sections": [
      {
        "expanded": false,
        "items": [
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Learn/Getting_started_with_the_web/JavaScript_basics",
            "slug": "Learn/Getting_started_with_the_web/JavaScript_basics",
            "title": "Основи JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Learn/JavaScript/First_steps",
            "slug": "Learn/JavaScript/First_steps",
            "title": "Перші кроки з JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Learn/JavaScript/Building_blocks",
            "slug": "Learn/JavaScript/Building_blocks",
            "title": "Конструктивні блоки JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Learn/JavaScript/Objects",
            "slug": "Learn/JavaScript/Objects",
            "title": "Вступ до об'єктів у JavaScript",
          },
        ],
        "title": "Для цілковитих початківців",
      },
      {
        "expanded": false,
        "items": [
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Introduction",
            "slug": "Web/JavaScript/Guide/Introduction",
            "title": "Вступ",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Grammar_and_Types",
            "slug": "Web/JavaScript/Guide/Grammar_and_Types",
            "title": "Граматика та типи",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Control_flow_and_error_handling",
            "slug": "Web/JavaScript/Guide/Control_flow_and_error_handling",
            "title": "Керування потоком та обробка помилок",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Loops_and_iteration",
            "slug": "Web/JavaScript/Guide/Loops_and_iteration",
            "title": "Цикли та ітерування",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Functions",
            "slug": "Web/JavaScript/Guide/Functions",
            "title": "Функції",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Expressions_and_Operators",
            "slug": "Web/JavaScript/Guide/Expressions_and_Operators",
            "title": "Вирази та оператори",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Numbers_and_dates",
            "slug": "Web/JavaScript/Guide/Numbers_and_dates",
            "title": "Числа та дати",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Text_formatting",
            "slug": "Web/JavaScript/Guide/Text_formatting",
            "title": "Форматування тексту",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Regular_Expressions",
            "slug": "Web/JavaScript/Guide/Regular_Expressions",
            "title": "Регулярні вирази",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Indexed_collections",
            "slug": "Web/JavaScript/Guide/Indexed_collections",
            "title": "Колекції з індексами",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Keyed_collections",
            "slug": "Web/JavaScript/Guide/Keyed_collections",
            "title": "Колекції з ключами",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Working_with_Objects",
            "slug": "Web/JavaScript/Guide/Working_with_Objects",
            "title": "Робота з об'єктами",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Details_of_the_Object_Model",
            "slug": "Web/JavaScript/Guide/Details_of_the_Object_Model",
            "title": "Докладно про об'єктну модель",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Using_promises",
            "slug": "Web/JavaScript/Guide/Using_promises",
            "title": "Використання промісів",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Iterators_and_generators",
            "slug": "Web/JavaScript/Guide/Iterators_and_generators",
            "title": "Ітератори та генератори",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Meta_programming",
            "slug": "Web/JavaScript/Guide/Meta_programming",
            "title": "Метапрограмування",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Guide/Modules",
            "slug": "Web/JavaScript/Guide/Modules",
            "title": "Модулі JavaScript",
          },
        ],
        "title": "Посібники з JavaScript",
      },
      {
        "expanded": false,
        "items": [
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Learn/Tools_and_testing/Client-side_JavaScript_frameworks",
            "slug": "Learn/Tools_and_testing/Client-side_JavaScript_frameworks",
            "title": "Клієнтські фреймворки у JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Learn/JavaScript/Client-side_web_APIs",
            "slug": "Learn/JavaScript/Client-side_web_APIs",
            "title": "Клієнтські API Вебу",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/A_re-introduction_to_JavaScript",
            "slug": "Web/JavaScript/A_re-introduction_to_JavaScript",
            "title": "Повторний вступ до JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Data_structures",
            "slug": "Web/JavaScript/Data_structures",
            "title": "Структури даних у JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Equality_comparisons_and_sameness",
            "slug": "Web/JavaScript/Equality_comparisons_and_sameness",
            "title": "Порівняння на схожість і однаковість",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Closures",
            "slug": "Web/JavaScript/Closures",
            "title": "Замикання",
          },
        ],
        "title": "Середній рівень",
      },
      {
        "expanded": false,
        "items": [
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Inheritance_and_the_prototype_chain",
            "slug": "Web/JavaScript/Inheritance_and_the_prototype_chain",
            "title": "Наслідування та ланцюжок прототипів",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Strict_mode",
            "slug": "Web/JavaScript/Reference/Strict_mode",
            "title": "Суворий режим",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Typed_arrays",
            "slug": "Web/JavaScript/Typed_arrays",
            "title": "Типізовані масиви JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Memory_Management",
            "slug": "Web/JavaScript/Memory_Management",
            "title": "Керування пам'яттю",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/EventLoop",
            "slug": "Web/JavaScript/EventLoop",
            "title": "Модель багатопотоковості та Цикл подій",
          },
        ],
        "title": "Поглиблене вивчення",
      },
    ],
  },
  {
    "links": [
      {
        "hasTranslation": true,
        "isCurrent": false,
        "path": "/uk/docs/Web/JavaScript/Reference/",
        "title": "Довідник",
      },
    ],
    "sections": [
      {
        "expanded": false,
        "items": [],
        "title": "Вбудовані об'єкти",
      },
      {
        "expanded": false,
        "items": [],
        "title": "Вирази та оператори",
      },
      {
        "expanded": false,
        "items": [],
        "title": "Інструкції та оголошення",
      },
      {
        "expanded": false,
        "items": [],
        "title": "Функції",
      },
      {
        "expanded": false,
        "items": [],
        "title": "Класи",
      },
      {
        "expanded": false,
        "items": [],
        "title": "Помилки",
      },
      {
        "expanded": false,
        "items": [
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/JavaScript_technologies_overview/",
            "title": "Огляд JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Lexical_grammar/",
            "title": "Лексична граматика",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Data_structures/",
            "title": "Структури даних у JavaScript",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties/",
            "title": "Перелічуваність і належність властивостей",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Iteration_protocols/",
            "title": "Протоколи ітерування",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Strict_mode/",
            "title": "Суворий режим",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Strict_mode/Transitioning_to_strict_mode/",
            "title": "Перехід до суворого режиму",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Template_literals/",
            "title": "Шаблонні літерали",
          },
          {
            "hasTranslation": true,
            "isCurrent": false,
            "path": "/uk/docs/Web/JavaScript/Reference/Deprecated_and_obsolete_features/",
            "title": "Нерекомендовані можливості",
          },
        ],
        "title": "Різне",
      },
    ],
  },
]
`;
